### ISSUE TITLE
API: build/publish should by async

### ISSUE THREAD

**Issue by davidferguson** [lektor/lektor#706]
The build/publish step as performed from the GUI dialog can be a long running task. For my site, this can take 30+ minutes if a full rebuild is required.

This obviously shouldn't be a sync API call, as forcing the user to have a browser page open for 30 minutes while a HTTP request completes is insane. Even for smaller sites, it makes much more sense to have this as an async call to, say `/buildpublish-start/` which starts a background task, and then another `/buildpublish-status/` which reports the status of the build/publish.

I also think (as shown in the example above) that there should be a single API endpoint for this, rather than the two as it is at the moment. This would allow someone to click publish, and then shutdown their machine and go and have coffee, whilst the background task builds the site and publishes it.

I'm not overly sure what the best way of achieving this is. I've created a simple version of it working [in my fork of lektor](https://github.com/mathshistory/mathshistory-lektor/commit/69f5b5234be80dc81054fab2e1c47d570d534f3e) if you want to have a look at that, but I'd be interested in comments on what we thing the best way of doing this is. With the current admin API being a single event bus and not supporting multiple sessions, we don't need anything fancy like redis/celery (I think?). In my early version it doesn't report the current status too which is a shame - I couldn't figure out how to do that with `click`.

Thoughts? @nixjdm @runfalk

**Comment by runfalk** [comment]
I think splitting it into two routes makes sense. Maybe we could even go as far as providing a single [Server Sent Events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events) endpoint to communicate any event from the backend. Basically everything that is reported in the CLI could be provided there, like current file and build errors.

Another thing that concerns me is how edits during such a deploy should be handled. Should all pages be locked for editing? What if a page that has already been published gets updated? The current solution technically doesn't prevent it since you can open another tab, but I think this is something we should protect against even with an async solution.

Personally I use CI to deploy when I push to my Git repository. I think all larger projects should probably consider a similar solution.

**Comment by davidferguson** [comment]
>I think splitting it into two routes makes sense.

I agree there should be a separate `/build` and `/publish` endpoints. But I think the `/publish` one should *also* build it. Surely you don't want to be publishing an out of date build?

>Maybe we could even go as far as providing a single Server Sent Events endpoint to communicate any event from the backend.

I wasn't aware of Server Sent Events until I saw that's how the data from the `/publish` endpoint sends data back. It looks like a good way of doing this. Looking at the code, we could use `BufferReporter` for this I think?

> Another thing that concerns me is how edits during such a deploy should be handled.

From #458, "*It is designed in a way that makes it seem like it should be possible to run multiple admins on one page (useful for testing). That's not possible due to a shared event bus for dialogs.*" So from the Admin GUI side of things, there are already issues with multiple instances. I get that the API could be used in other places, but at the moment I'm more concerned about fixing the Publish button in the GUI which currently **does not work** for sites that don't have a near instant build. The Web GUI is a requirement for us, as the editors won't necessary have any technical background.
